File: Android/TUIKit/TUICallKit/tuicallkit/src/main/java/com/tencent/qcloud/tuikit/tuicallkit/extensions/recents/RecentCallsViewModel.java
package com.tencent.qcloud.tuikit.tuicallkit.extensions.recents;

import android.app.Application;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MutableLiveData;

import com.tencent.qcloud.tuikit.TUICommonDefine;
import com.tencent.qcloud.tuikit.tuicallengine.TUICallDefine;
import com.tencent.qcloud.tuikit.tuicallengine.TUICallDefine.CallRecords;
import com.tencent.qcloud.tuikit.tuicallengine.TUICallEngine;

import java.util.ArrayList;
import java.util.List;

public class RecentCallsViewModel extends AndroidViewModel {
    private static final String TAG = "RecentCallsViewModel";

    private final MutableLiveData<List<CallRecords>> mCallHistoryLiveData = new MutableLiveData<>();
    private final MutableLiveData<List<CallRecords>> mCallMissedLiveData  = new MutableLiveData<>();

    public RecentCallsViewModel(@NonNull Application application) {
        super(application);
        mCallHistoryLiveData.setValue(new ArrayList<>());
        mCallMissedLiveData.setValue(new ArrayList<>());
    }

    public MutableLiveData<List<CallRecords>> getCallHistoryList() {
        return mCallHistoryLiveData;
    }

    public MutableLiveData<List<CallRecords>> getCallMissedList() {
        return mCallMissedLiveData;
    }

    public void queryRecentCalls(TUICallDefine.RecentCallsFilter filter) {
        TUICallEngine.createInstance(getApplication()).queryRecentCalls(filter, new TUICommonDefine.ValueCallback() {
            @Override
            public void onSuccess(Object data) {
                if (data == null || !(data instanceof List)) {
                    return;
                }
                List<CallRecords> queryList = (List<CallRecords>) data;
                if (filter != null && CallRecords.Result.Missed.equals(filter.result)) {
                    List<CallRecords> missList = mCallMissedLiveData.getValue();
                    if (missList != null) {
                        missList.removeAll(queryList);
                        missList.addAll(queryList);
                    }
                    mCallMissedLiveData.setValue(missList);
                } else {
                    List<CallRecords> historyList = mCallHistoryLiveData.getValue();
                    if (historyList != null) {
                        historyList.removeAll(queryList);
                        historyList.addAll(queryList);
                    }
                    mCallHistoryLiveData.setValue(historyList);
                }
            }

            @Override
            public void onError(int errCode, String errMsg) {
            }
        });
    }

    public void deleteRecordCalls(List<CallRecords> list) {
        if (list == null || list.isEmpty()) {
            return;
        }

        ArrayList<CallRecords> missList = new ArrayList<>(mCallMissedLiveData.getValue());
        missList.removeAll(list);
        mCallMissedLiveData.setValue(missList);

        List<CallRecords> allList = new ArrayList<>(mCallHistoryLiveData.getValue());
        allList.removeAll(list);
        mCallHistoryLiveData.setValue(allList);

        List<String> callIdList = new ArrayList<>();
        for (CallRecords record : list) {
            if (record != null && !TextUtils.isEmpty(record.callId)) {
                callIdList.add(record.callId);
            }
        }
        TUICallEngine.createInstance(getApplication()).deleteRecordCalls(callIdList,
                new TUICommonDefine.ValueCallback() {
                    @Override
                    public void onSuccess(Object data) {
                    }

                    @Override
                    public void onError(int errCode, String errMsg) {
                    }
                });
    }
}

File: Android/TUIKit/TUICallKit/tuicallkit-kt/src/main/java/com/tencent/qcloud/tuikit/tuicallkit/extensions/recents/RecentCallsViewModel.kt
package com.tencent.qcloud.tuikit.tuicallkit.extensions.recents

import android.app.Application
import android.text.TextUtils
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import com.tencent.qcloud.tuikit.TUICommonDefine
import com.tencent.qcloud.tuikit.tuicallengine.TUICallDefine.CallRecords
import com.tencent.qcloud.tuikit.tuicallengine.TUICallDefine.RecentCallsFilter
import com.tencent.qcloud.tuikit.tuicallengine.TUICallEngine

class RecentCallsViewModel(application: Application) : AndroidViewModel(application) {
    val callHistoryList = MutableLiveData<MutableList<CallRecords>?>()
    val callMissedList = MutableLiveData<MutableList<CallRecords>?>()

    init {
        callHistoryList.value = ArrayList()
        callMissedList.value = ArrayList()
    }

    fun queryRecentCalls(filter: RecentCallsFilter?) {
        TUICallEngine.createInstance(getApplication())
            .queryRecentCalls(filter, object : TUICommonDefine.ValueCallback<Any?> {
                override fun onSuccess(data: Any?) {
                    if (data == null || data !is List<*>) {
                        return
                    }
                    val queryList = data as List<CallRecords>
                    if (filter != null && CallRecords.Result.Missed == filter.result) {
                        val missList = callMissedList.value
                        if (missList != null) {
                            missList.removeAll(queryList)
                            missList.addAll(queryList)
                        }
                        callMissedList.setValue(missList)
                    } else {
                        val historyList = callHistoryList.value
                        if (historyList != null) {
                            historyList.removeAll(queryList)
                            historyList.addAll(queryList)
                        }
                        callHistoryList.setValue(historyList)
                    }
                }

                override fun onError(errCode: Int, errMsg: String) {}
            })
    }

    fun deleteRecordCalls(list: List<CallRecords>?) {
        if (list.isNullOrEmpty()) {
            return
        }
        val missList = ArrayList(callMissedList.value)
        missList.removeAll(list)
        callMissedList.value = missList
        val allList: MutableList<CallRecords> = ArrayList(callHistoryList.value)
        allList.removeAll(list)
        callHistoryList.value = allList
        val callIdList: MutableList<String> = ArrayList()
        for (record in list) {
            if (record != null && !TextUtils.isEmpty(record.callId)) {
                callIdList.add(record.callId)
            }
        }
        TUICallEngine.createInstance(getApplication()).deleteRecordCalls(callIdList,
            object : TUICommonDefine.ValueCallback<Any?> {
                override fun onSuccess(data: Any?) {}
                override fun onError(errCode: Int, errMsg: String) {}
            })
    }

    companion object {
        private const val TAG = "RecentCallsViewModel"
    }
}